File: ReiTunes.Server/Systemd.cs
using System.Diagnostics;
using CliWrap;
using Spectre.Console;

namespace ReiTunes.Server;

public static class Systemd
{
    /// <summary>
    /// Attempt to install the current process as a systemd service
    /// </summary>
    /// <returns>bool indicating whether the operation succeeded</returns>
    public static async Task<bool> InstallService(string? name = null, string description = "")
    {
        string processPath = Environment.ProcessPath!;
        string processFileName = Path.GetFileName(processPath);
        string serviceName = name ?? processFileName;

        AnsiConsole.WriteLine($"Installing {processFileName} as a systemd service...");

        string unitFileContents = @$"[Unit]
Description={description}

[Service]
Type=notify
ExecStart={processPath}
# fix for single-file apps with bundled native libraries
# https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file#including-native-libraries
Environment=""DOTNET_BUNDLE_EXTRACT_BASE_DIR=%h/.net""

[Install]
WantedBy=multi-user.target";

        string unitFilePath = $"/etc/systemd/system/{serviceName}.service";

        try
        {
            AnsiConsole.WriteLine("Writing unit file...");
            File.WriteAllText(unitFilePath, unitFileContents);
        }
        catch (Exception ex)
        {
            AnsiConsole.MarkupLine($"[red]Error:[/] write failed to {unitFilePath}. Did you forget to use sudo?");
            AnsiConsole.WriteException(ex.Demystify());
            return false;
        }

        try
        {
            AnsiConsole.WriteLine("Enabling service...");
            await (Cli.Wrap("systemctl").WithArguments($"enable {serviceName}") |
                (Console.WriteLine, Console.Error.WriteLine)).ExecuteAsync();

            AnsiConsole.WriteLine("Starting service...");
            await (Cli.Wrap("systemctl").WithArguments($"start {serviceName}") |
                (Console.WriteLine, Console.Error.WriteLine)).ExecuteAsync();
        }
        catch (Exception ex)
        {
            AnsiConsole.MarkupLine($"[red]Error:[/] failed to enable+start service");
            AnsiConsole.WriteException(ex.Demystify());
            return false;
        }

        AnsiConsole.MarkupLine($"[green]Done! Install succeeded.[/]");
        return true;
    }
}

File: reitunes-rs/src/systemd.rs
use anyhow::{Context, Result};
use std::env;
use std::fs;
use std::path::PathBuf;

pub fn install() -> Result<()> {
    let executable_path = env::current_exe().context("Failed to get current executable path")?;
    let service_name = "reitunes.service";
    let service_content = format!(
        r#"[Unit]
Description=ReITunes Music Player

[Service]
ExecStart={}
Restart=on-failure

[Install]
WantedBy=default.target
"#,
        executable_path.display()
    );

    let config_dir = dirs::config_dir()
        .context("Failed to get config directory")?
        .join("systemd")
        .join("user");

    fs::create_dir_all(&config_dir).context("Failed to create systemd user directory")?;

    let service_path: PathBuf = config_dir.join(service_name);
    fs::write(&service_path, service_content).context("Failed to write service file")?;

    println!("Systemd user service installed at: {}", service_path.display());
    println!("To start the service, run: systemctl --user start {}", service_name);
    println!("To enable the service to start on boot, run: systemctl --user enable {}", service_name);

    Ok(())
}
